### ISSUE TITLE
Missing bos and eos token on llama 3 sft training?

### ISSUE THREAD

**Issue by Nero10578** [axolotl-ai-cloud/axolotl#1608]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Is it expected that the bos and eos tokens <|begin_of_text|> and <|end_of_text> are supposed to be missing when running preprocess --debug?

### Current behaviour

Tokenized data when using parameters
special_tokens:
  eos_token: "<|eot_id|>"
  pad_token: "<|end_of_text|>"

I added the eos token following what was mentioned here https://github.com/OpenAccess-AI-Collective/axolotl/pull/1553#issuecomment-2103240958

```
[2024-05-11 05:14:06,672] [INFO] [axolotl.check_example_labels:45] [PID:322366] [RANK:0] <|start_header_id|>(-100, 128006) user(-100, 882) <|end_header_id|>(-100, 128007)

(-100, 271) A(-100, 32)  store(-100, 3637)  sells(-100, 31878)  two(-100, 1403)  types(-100, 4595)  of(-100, 315)  products(-100, 3956) :(-100, 25)  Product(-100, 5761)  X(-100, 1630)  and(-100, 323)  Product(-100, 5761)  Y(-100, 816) .(-100, 13)  The(-100, 578)  store(-100, 3637)  has(-100, 706)  a(-100, 264)  total(-100, 2860)  inventory(-100, 15808)  worth(-100, 5922)  $(-100, 400) 20(-100, 508) ,(-100, 11) 000(-100, 931) .(-100, 13)  If(-100, 1442)  Product(-100, 5761)  X(-100, 1630)  makes(-100, 3727)  up(-100, 709)  (-100, 220) 60(-100, 1399) %(-100, 4)  of(-100, 315)  the(-100, 279)  inventory(-100, 15808)  value(-100, 907)  and(-100, 323)  each(-100, 1855)  unit(-100, 5089)  of(-100, 315)  Product(-100, 5761)  X(-100, 1630)  costs(-100, 7194)  $(-100, 400) 40(-100, 1272) ,(-100, 11)  how(-100, 1268)  many(-100, 1690)  units(-100, 8316)  of(-100, 315)  Product(-100, 5761)  X(-100, 1630)  are(-100, 527)  in(-100, 304)  the(-100, 279)  inventory(-100, 15808) ?(-100, 30) <|eot_id|>(-100, 128009) <|start_header_id|>(-100, 128006) assistant(-100, 78191) <|end_header_id|>(-100, 128007)

(271, 271) To(1271, 1271)  find(1505, 1505)  the(279, 279)  number(1396, 1396)  of(315, 315)  units(8316, 8316)  of(315, 315)  Product(5761, 5761)  X(1630, 1630)  in(304, 304)  the(279, 279)  inventory(15808, 15808) ,(11, 11)  we(584, 584)  need(1205, 1205)  to(311, 311)  first(1176, 1176)  calculate(11294, 11294)  its(1202, 1202)  total(2860, 2860)  value(907, 907) .

(382, 382) Total(7749, 7749)  value(907, 907)  of(315, 315)  Product(5761, 5761)  X(1630, 1630)  =(284, 284)  $(400, 400) 20(508, 508) ,(11, 11) 000(931, 931)  x(865, 865)  (220, 220) 0(15, 15) .(13, 13) 6(21, 21)  =(284, 284)  $(400, 400) 12(717, 717) ,(11, 11) 000(931, 931)

(271, 271) Since(12834, 12834)  each(1855, 1855)  unit(5089, 5089)  of(315, 315)  Product(5761, 5761)  X(1630, 1630)  costs(7194, 7194)  $(400, 400) 40(1272, 1272) ,(11, 11)  we(584, 584)  can(649, 649)  divide(22497, 22497)  the(279, 279)  total(2860, 2860)  value(907, 907)  by(555, 555)  the(279, 279)  cost(2853, 2853)  per(824, 824)  unit(5089, 5089)  to(311, 311)  get(636, 636)  the(279, 279)  number(1396, 1396)  of(315, 315)  units(8316, 8316) :

(1473, 1473) $(3, 3) 12(717, 717) ,(11, 11) 000(931, 931)  �(1717, 1717) �(115, 115)  $(400, 400) 40(1272, 1272)  =(284, 284)  (220, 220) 300(3101, 3101) Therefore(55915, 55915) ,(11, 11)  there(1070, 1070)  are(527, 527)  (220, 220) 300(3101, 3101)  units(8316, 8316)  of(315, 315)  Product(5761, 5761)  X(1630, 1630)  in(304, 304)  the(279, 279)  inventory(15808, 15808) .(13, 13) <|eot_id|>(128009, 128009)
```

### Steps to reproduce

Process dataset with parameters:
    type: sharegpt
    conversation: llama-3

Running on Ubuntu WSL2 in Windows 10.

### Config yaml

```yaml
base_model: /home/owen/models/Meta-Llama-3-8B-Instruct
model_type: LlamaForCausalLM
tokenizer_type: AutoTokenizer
  
train_on_inputs: false
group_by_length: false
load_in_8bit: false
load_in_4bit: true
strict: false
sequence_len: 2048
bf16: true
fp16: false
tf32: false
flash_attention: true

# Data
datasets:
  - path: /home/owen/datasets/no-robots-sharegpt-fixed.jsonl
    type: sharegpt
    conversation: llama-3
 
warmup_steps: 10
dataset_prepared_path: ./last_run_prepared

# Iterations
num_epochs: 2
saves_per_epoch: 2

# Evaluation
val_set_size: 0.01
eval_table_size:
eval_table_max_new_tokens:
eval_sample_packing: false
evals_per_epoch: 4

# LoRA
output_dir: ./qlora-out
adapter: qlora
lora_model_dir:
lora_r: 64
lora_alpha: 64
lora_dropout: 0.05
lora_target_linear: true
lora_fan_in_fan_out:
lora_target_modules:
save_safetensors: true

# Sampling
sample_packing: true
pad_to_sequence_len: true

# Batching
gradient_accumulation_steps: 8
micro_batch_size: 4
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: true

# wandb
wandb_mode: # "offline" to save run metadata locally and not sync to the server, "disabled" to turn off wandb
wandb_project: llama-3-8b-instruct
wandb_entity: # A wandb Team name if using a Team
wandb_watch:
wandb_name: 64-64-2048-2ep-v0.5
wandb_run_id: # Set the ID of your wandb run
wandb_log_model: # "checkpoint" to log model to wandb Artifacts every `save_steps` or "end" to log only at the end of training

# Optimizer
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.0002

# Misc
early_stopping_patience:
resume_from_checkpoint:
logging_steps: 1
debug:
deepspeed: /axolotl/deepspeed_configs/zero3_bf16.json
weight_decay: 0.1
special_tokens:
  eos_token: "<|eot_id|>"
  pad_token: "<|end_of_text|>"
```

### Possible solution

I noticed that if I set the special tokens in the config yml, any changes in the fastchat monkeypatch does not make a difference to the tokenized dataset result. Its only if I don't set special tokens that changes in the monkeypatch would even do anything. Is this expected behaviour?

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [X] Windows

### Python Version

3.11.9

### axolotl branch-commit

b32c08f8cc1d21ef52eb80d424cb7f16e7e0f5b5

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by Nero10578** [comment]
Understood that the bos and eos token are not supposed to show up in the debug log.
